python/ray/serve/backend_state.py

```diff
@@ -46,87 +169,76 @@ def __init__(self, controller_name: str, detached: bool,
-            BackendTag, ReplicaTag, ActorHandle]] = dict()
-        self.currently_stopping_replicas: Dict[asyncio.Future, Tuple[
-            BackendTag, ReplicaTag]] = dict()
-
-        # Checkpointed state.
-        self.backends: Dict[BackendTag, BackendInfo] = dict()
-        self.backend_replicas: Dict[BackendTag, Dict[
-            ReplicaTag, ActorHandle]] = defaultdict(dict)
+        self._replicas: Dict[BackendTag, Dict[ReplicaState, List[
+            BackendReplica]]] = defaultdict(lambda: defaultdict(list))
+        self._backend_metadata: Dict[BackendTag, BackendInfo] = dict()
+        self._target_replicas: Dict[BackendTag, int] = defaultdict(int)
         self.backend_goals: Dict[BackendTag, GoalId] = dict()
-        self.backend_replicas_to_start: Dict[BackendTag, List[
-            ReplicaTag]] = defaultdict(list)
-        self.backend_replicas_to_stop: Dict[BackendTag, List[Tuple[
-            ReplicaTag, Duration]]] = defaultdict(list)
-        self.backends_to_remove: List[BackendTag] = list()
+
+        # Un-Checkpointed state.
+        self.pending_goals: Dict[GoalId, asyncio.Event] = dict()
 
         checkpoint = self._kv_store.get(CHECKPOINT_KEY)
         if checkpoint is not None:
-            (self.backends, self.backend_replicas, self.backend_goals,
-             self.backend_replicas_to_start, self.backend_replicas_to_stop,
-             self.backend_to_remove,
-             pending_goal_ids) = pickle.loads(checkpoint)
+            (self._replicas, self._backend_metadata, self._target_replicas,
+             self.backend_goals, pending_goal_ids) = pickle.loads(checkpoint)
 
             for goal_id in pending_goal_ids:
                 self._goal_manager.create_goal(goal_id)
 
-            # Fetch actor handles for all backend replicas in the system.
-            # All of these backend_replicas are guaranteed to already exist
-            # because they would not be written to a checkpoint in
-            # self.backend_replicas until they were created.
-            for backend_tag, replica_dict in self.backend_replicas.items():
-                for replica_tag in replica_dict.keys():
-                    replica_name = format_actor_name(replica_tag,
-                                                     self._controller_name)
-                    self.backend_replicas[backend_tag][
-                        replica_tag] = ray.get_actor(replica_name)
-
         self._notify_backend_configs_changed()
         self._notify_replica_handles_changed()
 
     def _checkpoint(self) -> None:
         self._kv_store.put(
             CHECKPOINT_KEY,
-            pickle.dumps(
-                (self.backends, self.backend_replicas, self.backend_goals,
-                 self.backend_replicas_to_start, self.backend_replicas_to_stop,
-                 self.backends_to_remove,
-                 self._goal_manager.get_pending_goal_ids())))
+            pickle.dumps((self._replicas, self._backend_metadata,
```

Comment:
Ah I see that we're just pickling the replica classes directly -- can we please avoid pickling the actor handles and object refs? The way Ray handles that is a bit weird and we should avoid it if possible

Reply:
Hmm I can definitely do that. I could do either: custom `__get_state__` and `__set_state__` or loop through every state and call something like `serialize()` and `deserialize()`. I personally prefer `__get_state__` and __set_state__`.

Reply:
get_state set_state SGTM